### PR TITLE
Migrate Pipeline Configuration to YAML

### PR DESCRIPTION
**Pull request by markysoft**
https://eaflood.atlassian.net/browse/FPD-285

The current pipeline is not under source control, being defined is the Azure DevOps GUI.
The functionality will implemented as a source controlled YAML file, with additional feature to provide:

- unit tests to be run during the build pipeline
- test results to be published
- containers and published to the registry on release
- containers will be tagged with the PR number or release version
- PRs will be deployed to an end point for review
- PR deployments and containers will be deleted on merge
- Helm charts published to the registry on code release

### CODE REVIEW

**Comment by markysoft** [comment]
I've made some changes to the code as the tests aren't unit tests, requiring Redis to be available. To get them running in the CI pipeline I've switched the 'catbox-redis' for 'catbox-memory' when running in the test environment.

I've ported across the PR workflow pipeline which determines the PR number, tags the container in the registry with the PR number, and deploys the code to a temporary endpoint for review. I've found this really useful during development to make sure it still runs after I've made changes.

At present we don't have CNAME wildcards setup, so the only way to view the PR deployment is to add a line to your C:\Windows\System32\drivers\etc\hosts file for the end point, i.e.
52.178.206.208 mine-support-pr8.defradev.com
This approach works well for web sites and APIs, but we will have to come up with a robust approach for services that consume from queues.
 
Once the PR is merged the PR container is removed from the registry, the deployments are cleaned up, and the temporary namespace deleted.
On a Github release the pipeline will tag the container with the version and push to the registry, along with a versioned Helm chart.

I've taken the opportunity to upgrade the node modules and do a bit of tidy up of file names too.

File: Dockerfile [hunk 1]
```diff
@@ -1,7 +1,7 @@
 FROM node:10.15.3-alpine
 
 # Create app directory
-WORKDIR /mine-support
+WORKDIR /code
```
[the remark this answers is not in the document]
Comment (markysoft) [reply Dockerfile]: It was for consistency across the projects, rather than having a folder name unique to the project.
The benfit for me is when mounting local folders in docker-compose it's more obvious what is being mounted.

We don't have in depth guidelines on docker files, but there a few notes in https://eaflood.atlassian.net/wiki/spaces/FFCPD/pages/1297219944/Standards+for+Development+Project+Structure+and+Naming+Conventions

If the team is happy with a move to having 'code' as the folder name, we could update the above page with the view to adding it to the docker standards of the developer guide https://defradev.visualstudio.com/DEFRA_FutureFarming/_git/defra-ff-documentation?path=%2Fdeveloper-guide%2Findex.md&version=GBmaster&createIfNew=true&_a=preview in future

[the remark this answers is not in the document]
Comment (johnwatson484) [reply Dockerfile]: I'm happy with code.  The less things we have to think about the better and agree it should go in the Developer guide.  There is a planned section for Docker, we can include it when we complete that ticket.

[the remark this answers is not in the document]
Comment (devotee101) [reply Dockerfile]: The convention seems to be /usr/src/app and I'd expect we'd follow that

I think we've got to also remember we're talking about creating folders on a Linux filesystem, so putting a folder under root is probably not something we should be doing given conventions in that space.

That said, /usr/src/app breaks Linux conventions, but it's broadly accepted (see https://github.com/nodejs/docker-node/issues/96)

TL;DR /code makes my eyes bleed!

[the remark this answers is not in the document]
Comment (johnwatson484) [reply Dockerfile]: Linked issue is interesting, seems some tools are dependent on us using that folder, could be something we need to use in the project is similar so we could avoid headaches later by following the convention.

As long as we're consistent and it's documented we still get the benefits @markysoft  has stated so I'm happy to go with the convention of /usr/src/app.

[the remark this answers is not in the document]
Comment (markysoft) [reply Dockerfile]: Seems to be a few thoughts on this, all different to our initial ones!
The official node.js sample uses `/usr/src/app`, with the caveats above, while others prefer `/opt`

I'm happy with the suggestions and the reasoning, and will change the WORKDIR to `/usr/src/app`
